### PR TITLE
feat: websocket subscription to entries

### PR DESCRIPTION
**Pull request by akhercha**
Resolves: #48 #61 

## Added
- `MODE` environment variable to be able to switch between production/dev,
- `/node/v1/data/subscribe` ws endpoint to be able to subscribe to certain pairs & get median,
- `perp_entries` & `future_entries` new tables

## Changed
- Updated rust nightly toolchain to most recent stable version (1.79.0) - needed for `aws-config` & `aws-sdk-secretsmanager`
- CI pragma-node port from `8080` to `3000` to fix CI test constantly failing

_____

Starkex Documentation:
https://docs.starkware.co/starkex/perpetual/becoming-an-oracle-provider-for-starkex.html

### CODE REVIEW

**Comment by akhercha** [comment]
@EvolveArt I updated to a more recent nightly toolchain but maybe we can use a stable version now? What was the reason for the nightly toolchain?

File: README.md
Comment (EvolveArt) [inline README.md:134]: probably `DEV_MODE=1` is more explicit

File: compose.dev.yaml
Comment (EvolveArt) [inline compose.dev.yaml:99]: revert

File: pragma-ingestor/src/main.rs
Comment (EvolveArt) [inline pragma-ingestor/src/main.rs:105]:
```suggestion
            "new perp entry created {} - {}({}) - {}",
```

File: pragma-node/src/handlers/entries/create_perp_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/create_perp_entry.rs:44]: so this message has to be exactly the same as the one constructed and signed in python. So in python/cairo we do not have such thing as `PerpEntry`.  But rather `FutureEntry` with an expiration timestamp. I think we should align on this implementation here as well.

File: pragma-node/src/handlers/entries/create_perp_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/create_perp_entry.rs:93]: extact this to a function `assert_publisher_active`

File: pragma-node/src/handlers/entries/create_perp_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/create_perp_entry.rs:23]: we can probably make this generic on some `Entry` trait

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/subscribe_to_entry.rs:108]: what happens when the consumer tries to subscribe to an unexisting pair ?

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/subscribe_to_entry.rs:181]: let's just use a generic `sign_data` function, as the params are given as args anyway. We can have a `sign_median_price_as_pragma` that would call the `sign_data` function

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/subscribe_to_entry.rs:191]: is the timestamp in ms or s ? let's call it `timestamp_ms` or `timestamp_s` for clarity

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/subscribe_to_entry.rs:211]: move these to `types.rs` file

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/subscribe_to_entry.rs:247]: move these to `utils.rs`

File: pragma-node/src/infra/repositories/entry_repository.rs
Comment (EvolveArt) [inline pragma-node/src/infra/repositories/entry_repository.rs:737]: comment to explain why use this method

File: pragma-node/src/infra/repositories/entry_repository.rs
Comment (EvolveArt) [inline pragma-node/src/infra/repositories/entry_repository.rs:866]: maybe we should use https://docs.timescale.com/api/latest/hyperfunctions/percentile-approximation/uddsketch/

File: pragma-node/src/infra/repositories/entry_repository.rs
Comment (EvolveArt) [inline pragma-node/src/infra/repositories/entry_repository.rs:893]: move to `constants.rs`

File: pragma-node/src/infra/repositories/entry_repository.rs
Comment (EvolveArt) [inline pragma-node/src/infra/repositories/entry_repository.rs:892]: let's not break the channel just return an empty response, or the latest result

File: pragma-node/src/main.rs
Comment (EvolveArt) [inline pragma-node/src/main.rs:115]: we could use a builder pattern here for simplicity and to make it easy to add different signer modes in the future. Let's create the issue and tackle it later

**Review by EvolveArt — CHANGES_REQUESTED** [review]
left few comments

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (akhercha) [reply pragma-node/src/handlers/entries/subscribe_to_entry.rs:108]: It's simply ignored - the server returns a `SubscriptionAck` messages that returns the currently subscribed pairs so the client can see which pairs successfully got registered

Comment (EvolveArt) [reply pragma-node/src/handlers/entries/subscribe_to_entry.rs:108]: ok lets not forget to add this in the docs

File: pragma-node/src/infra/repositories/entry_repository.rs
Comment (akhercha) [reply pragma-node/src/infra/repositories/entry_repository.rs:893]: I like the idea of having a constant.rs file for general things like `PRAGMA_ORACLE_NAME_IN_STARKEX` but do we want them for those kind of constants too?

I think it's clearer to have them very close to the function they're used in, make it clearer then to have every constants in one module even though they don't really share the same logic

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (akhercha) [reply pragma-node/src/handlers/entries/subscribe_to_entry.rs]: Couldn't move them there cause it created circular dependencies - moved them to the entry repository

File: README.md
Comment (akhercha) [reply README.md:134]: I like having `Mode` as an enum as it will allow us to allow more modes later by extending it (`staging`, `test` ...). what do you think? maybe a YAGNI case, no strong opinions

File: pragma-node/src/infra/repositories/entry_repository.rs
Comment (akhercha) [reply pragma-node/src/infra/repositories/entry_repository.rs:896]: Tried but this wasn't straightforward, errors & had a wrong median when it worked (probably bad implementation yeah). I'd say let's monitor & optimize if we have some bottleneck there

File: pragma-node/src/handlers/entries/create_perp_entry.rs
Comment (akhercha) [reply pragma-node/src/handlers/entries/create_perp_entry.rs]: done here:
https://github.com/astraly-labs/pragma-sdk/pull/109

File: pragma-node/src/handlers/entries/mod.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/mod.rs:226]:
```suggestion
    pub timestamp_s: UnixTimestamp,
```

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/subscribe_to_entry.rs:157]:
```suggestion
    message: Message,
```

we need to match on different message types, you can reuse what's done in other ws endpoint

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/subscribe_to_entry.rs:265]: I think we can make this whole code cleaner by leveraging a `Price` trait and having two structs `IndexPrice` and `MarkPrice` that implement it

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (EvolveArt) [inline pragma-node/src/handlers/entries/subscribe_to_entry.rs:376]: move to `utils/signing.rs` also we could leverage a `Signable` trait that we implement on the data types we want to be signable

File: pragma-node/src/infra/repositories/onchain_repository.rs
Comment (EvolveArt) [inline pragma-node/src/infra/repositories/onchain_repository.rs:28]: should we panic ?

File: pragma-node/src/utils/aws.rs
Comment (EvolveArt) [inline pragma-node/src/utils/aws.rs:41]: do we want to panic ?

**Review by EvolveArt — CHANGES_REQUESTED** [review]
let's remove perp entry too

File: pragma-node/src/utils/aws.rs
Comment (akhercha) [reply pragma-node/src/utils/aws.rs:41]: Depends on what we want, we can not panic but then it disable all our starkex endpoints. 

I like the fact that here we fail fast at the instanciation of the app - makes easier to fix it then to handle it everywhere later on (& monitor)

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (akhercha) [reply pragma-node/src/handlers/entries/subscribe_to_entry.rs:376]: Agreed - created an issue for this: #61

File: pragma-node/src/infra/repositories/onchain_repository.rs
Comment (akhercha) [reply pragma-node/src/infra/repositories/onchain_repository.rs]: nope! removed

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (akhercha) [reply pragma-node/src/handlers/entries/subscribe_to_entry.rs]: Yeah no, bad idea to stack too much TODOs.

I implemented it directly now

File: pragma-node/src/handlers/entries/subscribe_to_entry.rs
Comment (akhercha) [reply pragma-node/src/handlers/entries/subscribe_to_entry.rs]: done!

**Comment by akhercha** [comment]
Don't merge yet - I'm having a wrong price when dividing the mark price by the spot price of a stablecoin. Checking this

File: pragma-node/src/utils/signing/starkex.rs
Comment (EvolveArt) [inline pragma-node/src/utils/signing/starkex.rs:18]: why don't we use `self.pair_id` here ?

File: pragma-node/src/utils/signing/starkex.rs
Comment (EvolveArt) [inline pragma-node/src/utils/signing/starkex.rs:81]:
```suggestion
    fn try_get_hash(&self) -> Result<FieldElement, ConversionError> {
```

File: pragma-node/src/utils/signing/mod.rs
Comment (EvolveArt) [inline pragma-node/src/utils/signing/mod.rs:10]: use this_error trait

File: pragma-node/src/infra/repositories/entry_repository.rs
Comment (EvolveArt) [inline pragma-node/src/infra/repositories/entry_repository.rs:832]: no panic

**Review by EvolveArt — COMMENTED** [review]
last few comments before merging

File: pragma-node/src/utils/signing/starkex.rs
Comment (akhercha) [reply pragma-node/src/utils/signing/starkex.rs:18]: I'm also using the function elsewhere when encoding the asset id:
```rust
impl TryFrom<MedianEntryWithComponents> for AssetOraclePrice {
    type Error = ConversionError;

    fn try_from(median_entry: MedianEntryWithComponents) -> Result<Self, Self::Error> {
        let signed_prices: Result<Vec<SignedPublisherPrice>, ConversionError> = median_entry
            .components
            .into_iter()
            .map(SignedPublisherPrice::try_from)
            .collect();

        let global_asset_id = StarkexPrice::get_global_asset_id(&median_entry.pair_id)?;
```

**Review by EvolveArt — APPROVED** [review]
LGTM

**Pull request merged**
